dss/api/bundles/__init__.py

```diff
@@ -288,6 +262,76 @@ def delete(uuid: str, replica: str, json_request_body: dict, version: str = None
 
     return jsonify(response_body), status_code
 
+def build_bundle_file_metadata(replica: Replica, user_supplied_files: dict):
+    handle = Config.get_blobstore_handle(replica)
+
+    time_left = nestedcontext.inject("time_left")
+
+    # decode the list of files.
+    files = [{'user_supplied_metadata': file} for file in user_supplied_files]
+
+    while True:  # each time through the outer while-loop, we try to gather up all the file metadata.
+        for file in files:
+            user_supplied_metadata = file['user_supplied_metadata']
+            metadata_key = FileFQID(
+                uuid=user_supplied_metadata['uuid'],
+                version=user_supplied_metadata['version'],
+            ).to_key()
+            if 'file_metadata' not in file:
+                try:
+                    file_metadata = handle.get(replica.bucket, metadata_key)
+                except BlobNotFoundError:
+                    continue
+                file['file_metadata'] = json.loads(file_metadata)
+
+        # check to see if any file metadata is still not yet loaded.
+        for file in files:
```

Comment:
Move the files from the `files` list into a different list say `complete_files` when they have `file_metadata` added. Then you only need to check if `files` is true. It will also let you return a list of missing files at the end.

Reply:
This is a carryover of existing verification logic. I'd prefer to refactor in a separate PR.